trin-execution/src/era/binary_search.rs

```diff
@@ -0,0 +1,140 @@
+use e2store::{
+    e2store::types::{Entry, Header as E2StoreHeader},
+    era::{get_beacon_fork, CompressedSignedBeaconBlock, Era, SlotIndexStateEntry},
+    utils::get_era_file_download_links,
+};
+use surf::Client;
+
+use super::{
+    constants::FIRST_ERA_EPOCH_WITH_EXECUTION_PAYLOAD,
+    types::ProcessedEra,
+    utils::{download_raw_era, process_era_file},
+};
+
+pub struct EraBinarySearch {}
+
+impl EraBinarySearch {
+    /// For era files there isn't really a good way to know which era file has the block we want,
+    /// well we could download all of them and check that isn't really feasible. We could also
+    /// create a table which maps block numbers to era files, but that would be a lot of work and
+    /// would need to be updated every time a new era file is added. So we will just binary
+    /// search for the era file which contains the block we want. To make it so we don't have to
+    /// download multiple era files because they are very big we will download the first beacon
+    /// block estimate which blocks it contains and narrow our search till we find it.
+    pub async fn find_era_file(
+        http_client: Client,
+        block_number: u64,
+    ) -> anyhow::Result<ProcessedEra> {
+        let era_links = get_era_file_download_links(&http_client).await?;
+        let mut start_epoch_index = FIRST_ERA_EPOCH_WITH_EXECUTION_PAYLOAD;
+        let mut end_epoch_index = era_links[era_links.len() - 1]
+            .split('-')
+            .nth(1)
+            .expect("to be able to get epoch index")
+            .parse::<u64>()
+            .expect("to be able to parse epoch index");
+
+        while start_epoch_index <= end_epoch_index {
+            let mid = end_epoch_index + (start_epoch_index - end_epoch_index) / 2;
+            let mid_block = EraBinarySearch::download_first_beacon_block_from_era(
+                era_links[mid as usize].clone(),
+                http_client.clone(),
+            )
+            .await?;
+            let mid_block_number = mid_block.block.execution_block_number();
+
+            // minimize the worse case by fetching the first block number of the after mid .era
+            let mid_plus_one_block = EraBinarySearch::download_first_beacon_block_from_era(
```

Comment:
This will not work if epoch we are looking for is the last epoch, or later (epoch not yet created).

But I think that we can avoid looking for `mid_plus_one` every loop iteration. For example, we can simply check where `block_number` is in relationship to `mid_block_number` and `mid_block_number+8192`, something like this:
```rust
(start_epoch_index, end_epoch_index) = match block_number {
    ..=(mid_block_number - 1) => (start_epoch_index, mid - 1),
    (mid_block_number + 8192).. => (mid + 1, end_epoch_index),
    _ => {
        // mid is a likely candidate, but we are not sure
        // we should peek at the following epochs until we are sure
        let mut target = mid;
        while target < <last_epoch> {
            let next_block = download_first_beacon_block_from_era(target + 1);
            if next_block.block.execution_block_number() < block_number {
                // target is our epoch
                todo!("download and return target")
            } else {
                target += 1;
            }
        }
        // we are at the last epoch. we should download it and try it
        // note: this can probably be optimized by downloading the SlotIndexBlock and verifying,
        // but we can ignore that optimization
        todo!("download and try")
    },
}
```

This is just an example. I'm sure it can be made a bit more readable and maybe some functionality can be extracted.

Reply:
I realized that I have to work around the issue when `mid_block_number == 0` (first epoch with execution block). I think this can be as simple as
```rust
if mid_block_number == 0 {
    mid_block_number = MERGE_BLOCK;
}
```

Reply:
> This will not work if epoch we are looking for is the last epoch, or later (epoch not yet created).

For blocks from the last 6 months we should just be able to get them from the CL client. So if we are at the last epoch we can just through an error to say continue from CL client

Reply:
Returning error is fine. But it will panic with the current code when mid is equal to the last epoch (which is valid case). It will panic here:
```rust
let mid_plus_one_block = EraBinarySearch::download_first_beacon_block_from_era(
    era_links[mid as usize + 1].clone(),
    http_client.clone(),
)
```
because `mid as usize + 1` is non-existing index of `era_links`.

Reply:
I will add a if case to through an era in this case